hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/DirectoryDeletingService.java

```diff
@@ -222,12 +225,122 @@ public BackgroundTaskResult call() {
+          SnapshotInfo currSnapInfo = iterator.next().getValue();
+
+          // Expand deleted dirs only on active snapshot. Deleted Snapshots
+          // will be cleaned up by SnapshotDeletingService.
+          if (!currSnapInfo.getSnapshotStatus().equals(SNAPSHOT_ACTIVE) ||
+              currSnapInfo.getExpandedDeletedDir()) {
+            continue;
+          }
+
+          try (ReferenceCounted<IOmMetadataReader, SnapshotCache>
+                   rcCurrOmSnapshot = omSnapshotManager.checkForSnapshot(
+              currSnapInfo.getVolumeName(),
+              currSnapInfo.getBucketName(),
+              getSnapshotPrefix(currSnapInfo.getName()),
+              true)) {
+
+            OmSnapshot currOmSnapshot = (OmSnapshot) rcCurrOmSnapshot.get();
+            Table<String, OmKeyInfo> snapDeletedDirTable =
+                currOmSnapshot.getMetadataManager().getDeletedDirTable();
+
+            if (snapDeletedDirTable.isEmpty()) {
+              // TODO: [SNAPSHOT] Update Snapshot state using
+              //  SetSnapshotProperty from HDDS-7743 (YET TO BE MERGED)
+              continue;
+            }
+
+            Table.KeyValue<String, OmKeyInfo> deletedDirInfo;
+            List<Pair<String, OmKeyInfo>> allSubDirList
+                = new ArrayList<>((int) remainNum);
+
+            try (TableIterator<String, ? extends Table.KeyValue<String,
+                OmKeyInfo>> deletedIterator = snapDeletedDirTable.iterator()) {
+
+              long startTime = Time.monotonicNow();
+              deletedDirInfo = deletedIterator.next();
+
+              PurgePathRequest request = prepareDeleteDirRequest(
+                  remainNum, deletedDirInfo.getValue(),
+                  deletedDirInfo.getKey(), allSubDirList,
+                  currOmSnapshot.getKeyManager());
+              if (isBufferLimitCrossed(ratisByteLimit, consumedSize,
+                  request.getSerializedSize())) {
+                if (purgePathRequestList.size() != 0) {
+                  // if message buffer reaches max limit, avoid sending further
+                  remainNum = 0;
+                  break;
+                }
+                // if directory itself is having a lot of keys / files,
+                // reduce capacity to minimum level
+                remainNum = MIN_ERR_LIMIT_PER_TASK;
+                request = prepareDeleteDirRequest(
+                    remainNum, deletedDirInfo.getValue(),
+                    deletedDirInfo.getKey(), allSubDirList,
+                    currOmSnapshot.getKeyManager());
+              }
+
+              consumedSize += request.getSerializedSize();
+              purgePathRequestList.add(request);
+              remainNum = remainNum - request.getDeletedSubFilesCount();
+              remainNum = remainNum - request.getMarkDeletedSubDirsCount();
+              // Count up the purgeDeletedDir, subDirs and subFiles
+              if (request.getDeletedDir() != null
```

Comment:
nit: `getDeletedDir()` will produce not-null always if I'm not wrong.
```suggestion
              if (request.hasDeletedDir() &&
                  !request.getDeletedDir().isEmpty()) {
                dirNum++;
              }
```

Reply:
This is just a copy-pasta from the actual `DirectoryDeletingService` code. I think it should be fine.